### PR TITLE
[WIP] [FEAT lean-future] Event based data gathering

### PR DESCRIPTION
**Pull request by runspired**
Opening this as a tracking PR.

**Open Questions**

- Should HeimdallNode have a property responsible for "rollup" of stats for all leaves on the branch?
- how do annotations fit into the stats object structure?
- should we support JSON serialization of events/CounterStore? Or should we assume that someone that cares will implement transfer via binary streams?
- should we move configFor to HeimdallTree, as it isn't truly part of runtime needs?
- should we add `heimdall.startWithArgs` and if so, how would we add a dynamic schema to the CounterStore?

**TODO**

- [x] add time normalization step to leaf creation
- [x] add self time to leaves
- [x] reconstruct stats object from counter array and label graph
- [x] stats reconstruction tests
- [x] time reconstruction tests
- [x] HeimdallNode property for getting stats/annotations of top level leaves
- [x] HeimdallTree.constructFromJson() (for constructing from a previously serialized tree)
- [x] bump session identifier
- unit tests for shared code
  - [ ] CounterStore
  - [ ] EmptyObject
  - [ ] FastArray
  - [ ] HashMap
  - [ ] time
- [x] tests for heimdall runtime
- [ ] fix `Uint32Array.fill` for node 0.10 and 0.12

### CODE REVIEW

**Comment by hjdivad** [comment]
> Should HeimdallNode have a property responsible for "rollup" of stats for all leaves on the branch?

These go away during collection.  During analysis I think we'll want a cache for this kind of stuff (likely on the node) but I don't think we should build this yet; let's wait until we actually have a problem (ie we want to write queries against large trees and query time is slow).


> how do annotations fit into the stats object structure?

Haven't looked at the PR implementation yet but i suspect the answer is that they're orthogonal and the stats item is just an integer.  I'm wary of deciding to do that though; this feels like something that we need to benchmark.

> should we support JSON serialization of events/CounterStore? Or should we assume that someone that cares will implement transfer via binary streams?

Maybe later, and if so only if the cost of tree reification is enough that we care and we think that serializing the counter store & events is faster.

> should we move configFor to HeimdallTree, as it isn't truly part of runtime needs?

It's used by [ember-cli](https://github.com/ember-cli/ember-cli/blob/master/lib/cli/index.js#L47-L53).

Having said that its use requires that we know the stack.  It may be that we want to maintain a very simple stack to help with this kind of context.

> should we add heimdall.startWithArgs and if so, how would we add a dynamic schema to the CounterStore?

`startWithArgs` is probably the wrong name since `args` has a different meaning than in `setArgs`.  That said, good point about dynamic schemas.  Does indeed seem like an open question

File: README.md [hunk 1]
```diff
@@ -1,286 +1,99 @@
...
+ in which the overhead of a timer is enough to significantly alter stats.
+ 
+## Usage
+
+**instantiate**
 ```js
-var Heimdall = require('heimdalljs/heimdall');
-// this will create its own session and not use the global session
-var myInstance = new Heimdall();
+const heimdall = new Heimdall();
```
Comment (hjdivad) [inline README.md]: common case will be to use the global instance and not your own right?

File: README.md [hunk 2]
```diff
@@ -1,286 +1,99 @@
...
+```
 
-fs.mkdirSync = function () {
-  heimdall.statsFor('fs').mkdirCount++;
-  return origMkdirSync.apply(fs, arguments);
-}
+### Annotations
 
+```js
+heimdall.annotate(<annotation>);
```
Comment (hjdivad) [inline README.md]: i presume `annotate` is a better name for `setArgs`?

File: src/heimdall-tree/index.js [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+function statsFromCounters(counterStore, counterCache) {
+  if (!counterStore || !counterCache) {
+    return null;
+  }
+
+  return counterStore.restoreFromCache(counterCache);
+}
+
+export default class HeimdallTree {
+  constructor(heimdall) {
```
Comment (hjdivad) [inline src/heimdall-tree/index.js]: hmmm we should noodle through this.  I would expect us to be able to reuse this thing between in-process & parsing subsequent json files, but maybe not.

**Comment by hjdivad** [comment]
@runspired looks great.  Seems like you got pretty far on this!

File: src/shared/counter-store.js [hunk 4]
```diff
@@ -66,7 +66,11 @@ export default class CounterStore {
     this._config.push(numCounters);
 
     if (this._cache !== null) {
-      this._cache.push(NULL_NUMBER);
+      let cache = this._cache;
+
+      this._cache = new Uint32Array(this._namespaceCount);
```
Comment (stefanpenner) [inline src/shared/counter-store.js]: This isn't cross browser, we should fallback to an array

Comment (runspired) [inline src/shared/counter-store.js]: Which do we care about? It's IE10+. Older Node may have trouble, haven't fully investigated there.

**Comment by hjdivad** [comment]
Test failures are node 0.10 and 0.12.

Should deal with this for 0.3.x / 0.4.0 but we can ignore them for this

**Comment by stefanpenner** [comment]
I think the issue is just the UInt32Array usage

**Pull request merged**